Add chapter unpublishing confirmation modal GSoC M1.3

Kej-r03 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #16535.
2. This PR does the following: Add the UI for the chapter unpublishing confirmation modal as per [this mock](https://www.figma.com/file/mEmwgX2qSyVNpdKuNg1ZJC/Serial-Chapter-Launch-UI?type=design&node-id=12-14&t=NKqpmXb6rBcy4piP-4)

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->
The video below shows the feature of the unpublishing confirmation modal.

https://github.com/oppia/oppia/assets/99071926/5ece47a6-c31a-420c-ad4b-28a820aee250

The video below shows the responsiveness of the modal.


https://github.com/oppia/oppia/assets/99071926/dfcb76df-07b4-4564-9173-ec660529423a





#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

Kej-r03 committed:
Add chapter unpublishing confirmation modal

Rijuta-s commented:
@Kej-r03 Do we have a mock for mobile view? If not, I would suggest to have the warning section below the select reason section ( i.e stack them vertically). Mobile view looks very congested and difficult to read in my opinion.

Kej-r03 committed:
Change unpublish modal mobile view

Kej-r03 commented:
@Rijuta-s I have updated the mobile view



https://github.com/oppia/oppia/assets/99071926/7d87154e-3988-4e62-be52-d42c851b7488

Rijuta-s commented:
@Patel-Muhammad Can you please take a pass of the PR and make sure all the necessary code for making the component accessible is added and the color contrasts are right?

Rijuta-s commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:
```diff
@@ -1,18 +1,183 @@
+  <div class="modal-body">
+    <div class="reason-select">
+      <span>Select Reason:</span>
+
```
unnecessary empty line
```suggestion
```

Rijuta-s commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:
```diff
@@ -1,18 +1,183 @@
+    <div class="reason-select">
+      <span>Select Reason:</span>
+
+      <div ngbDropdown class="dropdown">
```
```suggestion
      <div ngbDropdown class="unpublishing-dropdown-menu">
```
Try to be more specific and unique with the class names. Ditto elsewhere

Rijuta-s commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:
```diff
@@ -1,18 +1,183 @@
+          {{ selectedReasonText }}
+        </div>
+        <div ngbDropdownMenu
+             class="dropdown-menu">
```
Ditto as above, it is difficult to  notice that `dropdown` and `dropdown-menu` are two different classes and are not applied to all dropdown menus and are specific to one dropdown menu only

Kej-r03 replied:
@Rijuta-s I have changed the class names. But they are kept similar to the corresponding angular dropdown directives.

Rijuta-s commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.ts:
```diff
@@ -18,21 +18,42 @@
+  selectedReasonText: string = this.badContentReasonText;
+  unpublishingReason: string = 'BAD_CONTENT';
+
+
```
```suggestion
```

Rijuta-s commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.ts:
```diff
@@ -18,21 +18,42 @@
+
+  splitChapterReasonText: string = 'Split Chapters' +
+    ' (requires new explorations to be added)';
+
```
```suggestion
```

Rijuta-s commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.ts:
```diff
@@ -18,21 +18,42 @@
 
+  badContentReasonText: string = 'Bad content (no new explorations ' +
+    'will be added)';
+
```
```suggestion
```

Rijuta-s reviewed (changes requested):
Left a few comments please take a look

Rijuta-s commented:
Also @Kej-r03 If this is a GSoC project PR it first needs to be approved by your mentor and then the reviewers of the PR. Thanks.

Nik-09 reviewed (commented):
Please update the PR description with the latest video (the one with the mobile view). For code reviews, I am aligning with the code owners.

Nik-09 reviewed (commented):
Please update the PR description with the latest video (the one with the mobile view). For code reviews, I am aligning with the code owners.

gp201 commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.spec.ts:
```diff
@@ -63,4 +79,18 @@ describe('Story Editor Unpublish Modal Component', () => {
+    mockPlatformFeatureService.
+      status.SerialChapterLaunchCurriculumAdminView.isEnabled = true;
+
+    component.setReason('BAD_CONTENT');
+    expect(component.unpublishingReason).toBe('BAD_CONTENT');
+    expect(component.selectedReasonText).toBe(component.badContentReasonText);
```
Add a pre-check for this test.

Kej-r03 replied:
Done

gp201 replied:
@Kej-r03 why are we setting the `unpublishingReason` as `BAD_CONTENT` if it is already set to `BAD_CONTENT`? Can we test with a different value before `setReason` is called

Kej-r03 replied:
@gp201 The default value of unpublishingReason is BAD_CONTENT before calling setReason(). If this test seems redundant maybe I can remove this unit test, and test for setReason('Bad_CONTENT') in the next unit test after checking 'CHAPTER_NEEDS_SPLITTING'. Does that sound fine?

gp201 replied:
@Kej-r03 Instead, can you set the `unpublishingReason` to any other value before executing `component.setReason('BAD_CONTENT');`?

gp201 replied:
@Kej-r03 Regarding this [comment](https://github.com/oppia/oppia/pull/18336#issuecomment-1591775837). We write the tests from a user perspective. Ideally, it would be better to have a placeholder. But let's test the scenario where we check if bad content is the default value as the test. Ignore the above comment and re-write the test.

gp201 commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.spec.ts:
```diff
@@ -63,4 +79,18 @@ describe('Story Editor Unpublish Modal Component', () => {
     component.confirm();
     expect(confirmSpy).toHaveBeenCalled();
   });
+
+  it('should set unpublishing reason', () => {
+    mockPlatformFeatureService.
+      status.SerialChapterLaunchCurriculumAdminView.isEnabled = true;
+
+    component.setReason('BAD_CONTENT');
+    expect(component.unpublishingReason).toBe('BAD_CONTENT');
+    expect(component.selectedReasonText).toBe(component.badContentReasonText);
+
+    component.setReason('CHAPTER_NEEDS_SPLITTING');
+    expect(component.unpublishingReason).toBe('CHAPTER_NEEDS_SPLITTING');
+    expect(component.selectedReasonText).toBe(
+      component.splitChapterReasonText);
```
Move this into a new unit test

Kej-r03 replied:
Done

gp201 reviewed (changes requested):
Thanks, @Kej-r03 ! I left some comments. PTAL

Kej-r03 committed:
Merge branch 'develop' into unpublish-modal

Kej-r03 committed:
Modify css class names and spec tests

Kej-r03 commented:
@Rijuta-s Some of the blank lines that you suggested to delete have to be kept, else they are showing linting errors.

Rijuta-s commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:
```diff
@@ -1,18 +1,182 @@
+  }
+
+  .modal-buttons .confirm-reason-button {
+    background-color: #f1dd9599;
```
Please use a hex color code or rgb(a) color code. Use other properties like opacity to get the required color. Here and elsewhere.

Kej-r03 replied:
Done

Rijuta-s commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:
```diff
@@ -1,18 +1,182 @@
+  }
+
+  .modal-buttons .cancel-button {
+    background-color: #79c979b8;
```
Please use a hex color code or rgb(a) color code. Use other properties like opacity to get the required color. Here and elsewhere.

Kej-r03 replied:
Done

Rijuta-s commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:
```diff
@@ -1,18 +1,182 @@
+  <div class="modal-body">
+    <div class="reason-select">
+      <span>Select Reason:</span>
+      <div ngbDropdown class="unpublishing-dropdown">
```
```suggestion
      <div ngbDropdown class="oppia-unpublishing-dropdown">
```
If the class is oppia specific it has to be prefixed with `oppia-` . This helps to distinguish classes from third party css classes like bootstrap. Ditto here and elsewhere.

Kej-r03 replied:
Done. I have done this only for the dropdown class names. Remaining classes are simply named based on their purposes.

Rijuta-s replied:
Purpose is fine, however we add `oppia-` prefix to distinguish that it is a class defined by us not by any third party library. Please add the prefix accordingly.

Patel-Muhammad commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:
```diff
@@ -1,18 +1,182 @@
+      <div ngbDropdown class="unpublishing-dropdown">
+        <div ngbDropdownToggle
+             aria-label="Dropdown toggle to select unpublishing reason"
+             aria-haspopup="menu"
```
Add `aria-expanded="false"` to indicate to the user that the dropdown is closed by default

Kej-r03 replied:
Done

Patel-Muhammad commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:
```diff
@@ -1,18 +1,182 @@
+        <div ngbDropdownMenu
+             class="unpublishing-dropdown-menu">
+          <div ngbDropdownItem
+               (click)="setReason('BAD_CONTENT')"
```
- Add:  `tabindex="0"` for keyboard focus to go to the element

Kej-r03 replied:
Done

Patel-Muhammad replied:
Also can you please add: 
- `(keydown.enter)="setReason('BAD_CONTENT')"`
to also make is accessible when user is uisng the keyboard.

Kej-r03 replied:
Done

Patel-Muhammad commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:
```diff
@@ -1,18 +1,182 @@
+    <div class="warning-and-confirmation">
+      <div class="warning-container">
+        <span>Warning:</span>
+        <div class="bad-content-warning warning-text" *ngIf="unpublishingReason=='BAD_CONTENT'">
```
- Make the warning focusable using `tabindex="0"`
- Add  `role="alert"`

Kej-r03 replied:
Done

Patel-Muhammad commented:
Hii @Kej-r03 added few comments please take a look. Thanks!

Kej-r03 committed:
Make accessibility changes

Kej-r03 commented:
@Patel-Muhammad @Rijuta-s I have addressed your comments. PTAL.

Patel-Muhammad commented:
@Kej-r03 LGTM! Thanks.

Rijuta-s commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:
```diff
@@ -1,18 +1,193 @@
+      message before proceeding.
+    </p>
+  </div>
+  <div class="modal-body">
```
```suggestion
  <div class="oppia-modal-body">
```
ditto elsewhere

Kej-r03 replied:
Now I have put the prefix in all those classes that are not used by third party. However modal-body class name is used by third party, so I have not changed it.

Patel-Muhammad commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:
```diff
@@ -1,18 +1,193 @@
+          </div>
+          <hr>
+          <div ngbDropdownItem
+               (click)="setReason('CHAPTER_NEEDS_SPLITTING')"
```
Also can you please add: 
- `(keydown.enter)="setReason('CHAPTER_NEEDS_SPLITTING')"`

to also make is accessible when user is uisng the keyboard.

Kej-r03 replied:
Done

Kej-r03 committed:
Change css class names

Kej-r03 commented:
@Rijuta-s @Patel-Muhammad I have made the changes. PTAL.

Rijuta-s reviewed (approved):
Thanks! Looks good to me

Patel-Muhammad commented:
@Kej-r03 changes LGTM. Thanks!

Kej-r03 commented:
@gp201 I had left a comment to your remark. Please take a look. Thanks

Kej-r03 committed:
Merge branch 'develop' into unpublish-modal

gp201 commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.ts:
```diff
@@ -18,21 +18,41 @@
+  splitChapterReasonText: string = 'Split Chapters' +
+    ' (requires new explorations to be added)';
+
+  selectedReasonText: string = this.badContentReasonText;
+  unpublishingReason: string = 'BAD_CONTENT';
```
Do we want `BAD_CONTENT` to be the default value? Or would it be better to have a placeholder such as `<Please select a reason>`?

Kej-r03 replied:
@gp201 As per the mock I feel it is better to keep the default value (It is not mentioned in the PRD to keep a placeholder by default), and also there will be nothing to display in the warning container if there is no reason selected, which does not look good as UI.

gp201 replied:
I will defer this decision to your mentor.

Kej-r03 commented:
@gp201 That would mean setting the reason to 'CHAPTER_NEEDS_SPLITTING' first (using `component.setReason('CHAPTER_NEEDS_SPLITTING')` and then executing `component.setReason('BAD_CONTENT');`. If that's the case, then I don't think we need 2 separate unit tests for 'CHAPTER_NNEDS_SPLITTING' and 'BAD_CONTENT', we can merge them into 1 test only, like I had done at the very start.

Kej-r03 committed:
Merge branch 'develop' into unpublish-modal

Kej-r03 commented:
> @gp201 That would mean setting the reason to 'CHAPTER_NEEDS_SPLITTING' first (using `component.setReason('CHAPTER_NEEDS_SPLITTING')` and then executing `component.setReason('BAD_CONTENT');`. If that's the case, then I don't think we need 2 separate unit tests for 'CHAPTER_NNEDS_SPLITTING' and 'BAD_CONTENT', we can merge them into 1 test only, like I had done at the very start.

Hi @gp201 Can you please suggest what needs to be done about this.
Also can you suggest how can I pass the failing tests in the PR.

gp201 commented:
@Kej-r03 I have responded to your comment [here](https://github.com/oppia/oppia/pull/18336/files#r1233099218) PTAL.

Regarding the e2e failure looks like the following error is occurring:

```console
Error: core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:1:13 - error TS2341: Property 'platformFeatureService' is private and only accessible within class 'StoryEditorUnpublishModalComponent'.

1 <div *ngIf="platformFeatureService.status.SerialChapterLaunchCurriculumAdminView.isEnabled">
              ~~~~~~~~~~~~~~~~~~~~~~

  core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.ts:25:16
    25   templateUrl: './story-editor-unpublish-modal.component.html'
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component StoryEditorUnpublishModalComponent.


Error: core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.html:72:14 - error TS2341: Property 'platformFeatureService' is private and only accessible within class 'StoryEditorUnpublishModalComponent'.

72 <div *ngIf="!platformFeatureService.status.SerialChapterLaunchCurriculumAdminView.isEnabled">
                ~~~~~~~~~~~~~~~~~~~~~~

  core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.ts:25:16
    25   templateUrl: './story-editor-unpublish-modal.component.html'
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component StoryEditorUnpublishModalComponent.

```

Kej-r03 committed:
Merge branch 'develop' into unpublish-modal

Kej-r03 committed:
Rectify frontend tests

Kej-r03 commented:
@gp201 I have updated the test. Now it checks the initial value to be BAD_CONTENT, then it checks after changing the value to CHAPTER_NEEDS_SPLITTING, then again changes the reason to BAD_CONTENT. I think this covers the user journey.

Also for the failing test, I have now added a function for getting the value of the feature flag. I hope it wont fail anymore.

gp201 commented on core/templates/pages/story-editor-page/modal-templates/story-editor-unpublish-modal.component.spec.ts:
```diff
@@ -63,4 +79,26 @@ describe('Story Editor Unpublish Modal Component', () => {
+    expect(component.isSerialChapterFeatureFlagEnabled()).toEqual(true);
+  });
+
+  it('should set unpublishing reason', () => {
+    expect(component.unpublishingReason).toBe('BAD_CONTENT');
+    expect(component.selectedReasonText).toBe(component.badContentReasonText);
+
+    component.setReason('CHAPTER_NEEDS_SPLITTING');
+    expect(component.unpublishingReason).toBe('CHAPTER_NEEDS_SPLITTING');
+    expect(component.selectedReasonText).toBe(
+      component.splitChapterReasonText);
+
+    component.setReason('BAD_CONTENT');
+    expect(component.unpublishingReason).toBe('BAD_CONTENT');
+    expect(component.selectedReasonText).toBe(component.badContentReasonText);
+  });
```
```suggestion
// checks if the default reason for publishing is bad content
  it('should ...', () => {
    expect(component.unpublishingReason).toBe('BAD_CONTENT');
    expect(component.selectedReasonText).toBe(component.badContentReasonText);
 }
  it('should set unpublishing reason', () => {
    expect(component.unpublishingReason).toBe('BAD_CONTENT');
    expect(component.selectedReasonText).toBe(component.badContentReasonText);

    component.setReason('CHAPTER_NEEDS_SPLITTING');

    expect(component.unpublishingReason).toBe('CHAPTER_NEEDS_SPLITTING');
    expect(component.selectedReasonText).toBe(
      component.splitChapterReasonText);
  });
```

Kej-r03 replied:
@gp201 Done. In the second test, I also added the test to check for setting the reason to BAD_CONTENT after CHAPTER_NEEDS_SPLITTING, else frontend coverage check is not passing.

gp201 reviewed (changes requested):
1 final nit

Kej-r03 committed:
Merge branch 'develop' into unpublish-modal

Kej-r03 committed:
Rectify frontend tests

gp201 reviewed (approved):
LGTM for codeowner files

Kej-r03 committed:
Merge branch 'develop' into unpublish-modal

Kej-r03 committed:
Merge branch 'develop' into unpublish-modal

Kej-r03 committed:
Merge branch 'develop' into unpublish-modal
